README.md
# FUNK OFF :chart_with_upwards_trend:

[![GitHub CI](https://github.com/iamtheblackunicorn/funkoff/actions/workflows/dart.yml/badge.svg)](https://github.com/iamtheblackunicorn/funkoff/actions)

*An over-powered engine to evaluate mathematical graph functions.* :chart_with_upwards_trend:

## About :books:

The other day I wanted to compute a number of values for a function formula. I wondered how I could implement a function that handles cubes, squares and linear functions fed into a code function as a string and produce a list of values as Ys from a given list of Xs. *Funk Off* is that library. It supports linear functions, quadratic functions, and cubic functions!

## Installation :inbox_tray:

### Adding to your project

To add *Funk Off* to your project's dependencies, add this line to your project's `pubspec.yaml`:

#### From *GitHub*

```YAML
depdencies:
  ...
  funkoff:
    git: git://github.com/iamtheblackunicorn/FunkOff.git
```

#### From *Pub.dev*

```YAML
depdencies:
  ...
  funkoff: ^1.4.0
```

## Usage :hammer:

### Importing

Import the engine-API like this:

```dart
import 'package:funkoff/funkoff.dart';
```

## Note :scroll:

- *Funk Off :chart_with_upwards_trend:* by Alexander Abraham :black_heart: a.k.a. *"The Black Unicorn" :unicorn:*
- Licensed under the MIT license.


CHANGELOG.md
# CHANGELOG

## Version 1.0.0: Initial release

- initial release to GitHub
- initial upload to Pub.dev

## Version 1.1.0: Functions shell

- added an API for evaluating functions and datasets in a shell
- added a shell binary

## Version 1.1.1: Scripting!

- added a scripting engine
- scripts with the right syntax and the extension `fs` can now be run with `funkoff`.

## Version 1.2.0-nullsafety.0: Null Safety

- made package null safe
- minor fixes

## Version 1.3.0-nullsafety.0: Milestone update

- updated docs
- minor fixes

## Version 1.4.0: The Milestone Release Update

- no longer pre-release


lib/shellAPI.dart
import 'dart:io';
import 'coreEngine.dart';
import 'cubeFunctions.dart';
import 'quadraticFunctions.dart';
import 'linearFunctions.dart';

/// This function evaluates yValues from given user input
/// of the form [formula=>dataset].
String evaluateIO(String userInput) {
  String result = 'NONE';
  try {
    String formula = userInput.split('=>')[0];
    List<String> stringDataSet = userInput.split('=>')[1].split(' ');
    List<double> dataSet = [];
    for (int i = 0; i < stringDataSet.length; i++) {
      dataSet.add(double.tryParse(stringDataSet[i]));
    }
    if (astParser(formula) == 'LINEAR_FUNCTION') {
      List<double> yValues = computeLinearSet(formula, dataSet);
      result = yValues.join(' ');
    } else if (astParser(formula) == 'QUADRATIC_FUNCTION') {
      List<double> yValues = computeQuadraticSet(formula, dataSet);
      result = yValues.join(' ');
    } else if (astParser(formula) == 'CUBE_FUNCTION') {
      List<double> yValues = computeCubeSet(formula, dataSet);
      result = yValues.join(' ');
    } else {}
  } catch (e) {
    result = 'NONE';
  }
  return result;
}

/// This function check whether the shell input is valid!
bool checkInput(String userInput) {
  bool result = false;
  if (evaluateIO(userInput) == 'NONE') {
  } else {
    result = true;
  }
  return result;
}

/// Provides a small shell for testing functions and datasets!
void functionsShell() {
  String userPrompt = 'shell';
  String actualPrompt = 'f(x)';
  while (userPrompt != 'quit') {
    stdout.write('$actualPrompt: ');
    userPrompt = stdin.readLineSync();
    if (userPrompt == 'quit') {
      exit(0);
    } else if (checkInput(userPrompt) == true) {
      print(evaluateIO(userPrompt));
    } else {
      print('Syntax error!');
    }
  }
}

/// This function processes a script with the ending *.fs.
void evaluateScript(String fileName) {
  RegExp fileRegex = RegExp(r'.*\.fs');
  if (fileRegex.hasMatch(fileName) == true) {
    List<String> contentList = File(fileName).readAsStringSync().split('\n');
    for (int i = 0; i < contentList.length; i++) {
      print(evaluateIO(contentList[i]));
    }
  } else {
    print('No valid file supplied!');
  }
}

/// Tests all of the above!
void testShellFunctions() {
  print(evaluateIO('300.0 + ( 24.0 * a )=>4.2 3.2 4.5 4.6'));
  print(evaluateIO('300.0 + ( 25.0 * t * t ) + ( 24.0 * t )=>4.2 3.2 4.5 4.6'));
  print(evaluateIO(
      '300.0 + ( 25.0 * t * t * t ) + ( 26.0 * t * t ) + ( 24.0 * t )=>4.2 3.2 4.5 4.6'));
  print(evaluateIO('=>'));

  print(checkInput('300.0 + ( 24.0 * a )=>4.2 3.2 4.5 4.6'));
  print(checkInput('300.0 + ( 25.0 * t * t ) + ( 24.0 * t )=>4.2 3.2 4.5 4.6'));
  print(checkInput(
      '300.0 + ( 25.0 * t * t * t ) + ( 26.0 * t * t ) + ( 24.0 * t )=>4.2 3.2 4.5 4.6'));
}


lib/core/coreEngine.dart
Map<String, dynamic> enginePatterns() {
  Map<String, dynamic> patterns = {
    'NUMBER': RegExp(r'[0-9]+\.[0-9]+'),
    'PLUS': RegExp(r'\+'),
    'MINUS': RegExp(r'\-'),
    'TIMES': RegExp(r'\*'),
    'DYNAMIC': RegExp(r'[a-z]'),
    'DIVIDE': RegExp(r'\/'),
    'OPEN_PAREN': RegExp(r'\('),
    'CLOSE_PAREN': RegExp(r'\)')
  };
  return patterns;
}

Map<String, dynamic> engineAST() {
  Map<String, dynamic> ast = {
    'LINEAR_FUNCTION':
        'NUMBER PLUS OPEN_PAREN NUMBER TIMES DYNAMIC CLOSE_PAREN',
    'QUADRATIC_FUNCTION':
        'NUMBER PLUS OPEN_PAREN NUMBER TIMES DYNAMIC TIMES DYNAMIC CLOSE_PAREN PLUS OPEN_PAREN NUMBER TIMES DYNAMIC CLOSE_PAREN',
    'CUBE_FUNCTION':
        'NUMBER PLUS OPEN_PAREN NUMBER TIMES DYNAMIC TIMES DYNAMIC TIMES DYNAMIC CLOSE_PAREN PLUS OPEN_PAREN NUMBER TIMES DYNAMIC TIMES DYNAMIC CLOSE_PAREN PLUS OPEN_PAREN NUMBER TIMES DYNAMIC CLOSE_PAREN'
  };
  return ast;
}

Map<String, dynamic> cosometicStatements() {
  Map<String, dynamic> cosmetics = {
    'PLUS': RegExp(r'\+'),
    'MINUS': RegExp(r'\-'),
    'TIMES': RegExp(r'\*'),
    'DIVIDE': RegExp(r'\/'),
    'OPEN_PAREN': RegExp(r'\('),
    'CLOSE_PAREN': RegExp(r'\)'),
    'DYNAMIC': RegExp(r'[a-z]')
  };
  return cosmetics;
}

List<dynamic> patternParser(String formula) {
  Map<String, dynamic> patterns = enginePatterns();
  List<dynamic> resultList = [];
  List<String> tokens = formula.split(' ');
  for (int i = 0; i < tokens.length; i++) {
    for (int x = 0; x < patterns.length; x++) {
      String key = patterns.keys.elementAt(x);
      if (patterns[key].hasMatch(tokens[i]) == true) {
        String match = patterns[key].stringMatch(tokens[i]);
        List<String> imList = [key, match];
        resultList.add(imList);
      } else {}
    }
  }
  return resultList;
}

String astParser(String formula) {
  Map<String, dynamic> ast = engineAST();
  List<dynamic> parsedTokens = patternParser(formula);
  List<String> tokenList = [];
  String result = 'NONE';
  for (int i = 0; i < parsedTokens.length; i++) {
    tokenList.add(parsedTokens[i][0]);
  }
  for (int x = 0; x < ast.length; x++) {
    String key = ast.keys.elementAt(x);
    if (ast[key] == tokenList.join(' ')) {
      result = key;
      break;
    } else {}
  }
  return result;
}

bool statementLinter(String formula) {
  bool result = false;
  if (astParser(formula) != 'NONE') {
    result = true;
  } else {}
  return result;
}

List<double> irParser(String formula) {
  List<double> result = [];
  if (astParser(formula) != 'NONE') {
    List<dynamic> parsedTokens = patternParser(formula);
    Map<String, dynamic> cosmetics = cosometicStatements();
    for (int i = 0; i < parsedTokens.length; i++) {
      for (int x = 0; x < cosmetics.length; x++) {
        String key = cosmetics.keys.elementAt(x);
        if (cosmetics[key].hasMatch(parsedTokens[i][1])) {
          parsedTokens.remove(parsedTokens[i]);
        } else {}
      }
    }
    for (int y = 0; y < parsedTokens.length; y++) {
      if (RegExp(r'[0-9]+\.[0-9]+').hasMatch(parsedTokens[y][1]) == true) {
        result.add(double.parse(parsedTokens[y][1]));
      } else {}
    }
  } else {}
  return result;
}

void makeVisualTree(String formula) {
  String spaces = '   ';
  print('\n');
  print('\n');
  print(formula);
  print('$spaces|');
  print('$spaces|');
  for (int i = 0; i < patternParser(formula).length; i++) {
    String tokenType = patternParser(formula)[i][0];
    String token = patternParser(formula)[i][1];
    print('$spaces|');
    print('$spaces--- $tokenType: $token');
  }
  print('\n');
  print('\n');
}

void testEngine() {
  String formula = '300.0 + ( 25.0 * t * t ) + ( 24.0 * t )';
  makeVisualTree(formula);
  print(patternParser(formula));
  print(astParser(formula));
}

void main() {
  testEngine();
}


lib/linearFunctions.dart
import 'funkoff.dart';

/// Computes Y from a given X and function formula
/// and returns it as a double for linear functions.
double computeLinear(String formula, double value) {
  double result = 0;
  if (astParser(formula) == 'LINEAR_FUNCTION') {
    double constant = irParser(formula)[0];
    double factor = irParser(formula)[1];
    result = constant + (factor * value);
  } else {}
  return result;
}

/// Computes Ys from a given Xs and function formula
/// and returns it as a List of doubles for linear functions.
List<double> computeLinearSet(String formula, List<double> values) {
  List<double> results = [];
  for (int i = 0; i < values.length; i++) {
    double result = computeLinear(formula, values[i]);
    results.add(result);
  }
  return results;
}

/// Tests the above methods for linear functions.
void testLinears() {
  String testFormulaLinear = '300.0 + ( 25.0 * t )';
  List<double> testSet = [4.5, 4.6, 4.7, 4.8];
  print(patternParser(testFormulaLinear));
  print(astParser(testFormulaLinear));
  print(irParser(testFormulaLinear));
  print(computeLinear(testFormulaLinear, 2));
  print(computeLinearSet(testFormulaLinear, testSet));
}


lib/src/cubeFunctions.dart
import 'coreEngine.dart';
import 'dart:math';

/// Computes Y from a given X and function formula
/// and returns it as a double for cubic functions.
double computeCube(String formula, double value) {
  double result = 0;
  if (astParser(formula) == 'CUBE_FUNCTION') {
    double constant = irParser(formula)[0];
    double factorToBeCubed = irParser(formula)[1];
    double factorToBeSquared = irParser(formula)[2];
    double factorToBeMultiplied = irParser(formula)[3];
    result = constant +
        (factorToBeCubed * pow(value, 3)) +
        (factorToBeSquared * pow(value, 2)) +
        (factorToBeMultiplied * value);
  } else {}
  return result;
}

/// Computes Ys from a given Xs and function formula
/// and returns it as a List of doubles for cubic functions.
List<double> computeCubeSet(String formula, List<double> values) {
  List<double> results = [];
  for (int i = 0; i < values.length; i++) {
    double result = computeCube(formula, values[i]);
    results.add(result);
  }
  return results;
}

/// Tests the above methods for cubic functions.
void testCubes() {
  String testFormulaCube =
      '300.0 + ( 25.0 * t * t * t ) + ( 26.0 * t * t ) + ( 24.0 * t )';
  List<double> testSet = [4.5, 4.6, 4.7, 4.8];
  print(patternParser(testFormulaCube));
  print(astParser(testFormulaCube));
  print(irParser(testFormulaCube));
  print(computeCube(testFormulaCube, 2));
  print(computeCubeSet(testFormulaCube, testSet));
}


lib/quadraticFunctions.dart
import 'coreEngine.dart';
import 'dart:math';

/// Computes Y from a given X and function formula
/// and returns it as a double for quadratic functions.
double computeQuadratic(String formula, double value) {
  double result = 0;
  if (astParser(formula) == 'QUADRATIC_FUNCTION') {
    double constant = irParser(formula)[0];
    double factorToBeSquared = irParser(formula)[1];
    double factorToBeMultiplied = irParser(formula)[2];
    result = constant +
        (factorToBeSquared * pow(value, 2)) +
        (factorToBeMultiplied * value);
  } else {}
  return result;
}

/// Computes Ys from a given Xs and function formula
/// and returns it as a List of doubles for quadratic functions.
List<double> computeQuadraticSet(String formula, List<double> values) {
  List<double> results = [];
  for (int i = 0; i < values.length; i++) {
    double result = computeQuadratic(formula, values[i]);
    results.add(result);
  }
  return results;
}

/// Tests the above methods for quadratic functions.
void testQuadratics() {
  String testFormulaQuadratic = '300.0 + ( 25.0 * t * t ) + ( 24.0 * t )';
  List<double> testSet = [4.5, 4.6, 4.7, 4.8];
  print(patternParser(testFormulaQuadratic));
  print(astParser(testFormulaQuadratic));
  print(irParser(testFormulaQuadratic));
  print(computeQuadratic(testFormulaQuadratic, 2));
  print(computeQuadraticSet(testFormulaQuadratic, testSet));
}


example/lib/main.dart
/*
FUNK OFF by Alexander Abraham a.k.a. "The Black Unicorn".
Licensed under the MIT license.
*/

import 'package:funkoff/funkoff.dart'; // If you want to use the engine itself, use this!

void main() {
  String linearFunction = '300.0 + ( 25.0 * t )';
  String quadraticFunction = '300.0 + ( 25.0 * t * t ) + ( 24.0 * t )';
  String cubeFunction =
      '300.0 + ( 25.0 * t * t * t ) + ( 26.0 * t * t ) + ( 24.0 * t )';
  List<double> testSet = [4.5, 4.6, 4.7, 4.8];
  double testX = 2;

  print(patternParser(linearFunction));
  print(astParser(linearFunction));
  print(statementLinter(linearFunction));
  print(irParser(linearFunction));
  makeVisualTree(linearFunction);

  print(patternParser(quadraticFunction));
  print(astParser(quadraticFunction));
  print(statementLinter(quadraticFunction));
  print(irParser(quadraticFunction));
  makeVisualTree(quadraticFunction);

  print(patternParser(cubeFunction));
  print(astParser(cubeFunction));
  print(statementLinter(cubeFunction));
  print(irParser(cubeFunction));
  makeVisualTree(cubeFunction);

  print(computeLinear(linearFunction, testX));
  print(computeQuadratic(quadraticFunction, testX));
  print(computeCube(cubeFunction, testX));

  print(computeLinearSet(linearFunction, testSet));
  print(computeQuadraticSet(quadraticFunction, testSet));
  print(computeCubeSet(cubeFunction, testSet));

  testEngine();
  testCubes();
  testQuadratics();
  testLinears();
}
